Handle NotImplemented from __ne__

kamahen opened the pull request:
Also change sharedparse.py to avoid this kind of bug in future.
Fixes #4767.

JelleZijlstra commented on test-data/unit/check-warnings.test:
```diff
@@ -145,8 +145,29 @@ main:4: warning: Returning Any from function declared to return "int"
 class A:
-    def __eq__(self, other: object) -> bool: return NotImplemented
+    def __init__(self, x: Any) -> None:
+      self.x = x
```
Indent with four spaces.

kamahen replied:
Done

JelleZijlstra commented on test-data/unit/check-warnings.test:
```diff
@@ -145,8 +145,29 @@ main:4: warning: Returning Any from function declared to return "int"
+      if self.__class__ is other.__class__:
+        return bool(self.x != other.x)
+      else:
+        # pylint will complain about this return not needing
```
This comment seems irrelevant; we don't run pylint on test cases.

kamahen replied:
The reason I put this comment in is that someone might think that the "else" could be removed, leaving the "return" outside then if-then-else (pylint encourages this kind of thing). Do you want me to just remove the comment about pylint complaining?

JelleZijlstra replied:
I don't think the comment is helpful, but I won't be the person deciding whether to merge this.

Why does it matter whether the `else` is there? It seems to me that it shouldn't matter to mypy, since the control flow should be the same in both cases.

kamahen replied:
The type inferencing might produce different results (I'm essentially doing black-box test thinking). So, I have two ways of generating a Union result.

JelleZijlstra commented on test-data/unit/check-warnings.test:
```diff
@@ -145,8 +145,29 @@ main:4: warning: Returning Any from function declared to return "int"
+        # works as the inferred return typ
+        return NotImplemented
+    # TODO: If the programmer specifies "-> Union[bool, type(NotImplemented)]",
+    #       mypy suggests changing this to "-> Union[bool, type[NotImplemented]]",
```
Does `Type[NotImplemented]` work? Maybe we should suggest that.

kamahen replied:
```
>>> from typing import *
>>> Type[NotImplemented]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.5/typing.py", line 996, in __getitem__
    params = tuple(_type_check(p, msg) for p in params)
  File "/usr/lib/python3.5/typing.py", line 996, in <genexpr>
    params = tuple(_type_check(p, msg) for p in params)
  File "/usr/lib/python3.5/typing.py", line 312, in _type_check
    raise TypeError(msg + " Got %.100r." % (arg,))
TypeError: Parameters to generic types must be types. Got NotImplemented.
```

gvanrossum replied:
That's a runtime error, and there's nothing you can do about it. The return type should just be `-> bool`.

kamahen commented on mypy/sharedparse.py:
```diff
@@ -74,13 +46,45 @@
+    "__getattribute__",
+    "__getitem__",
+    "__hex__",
+    "__init__",
```
@gvanrossum - I've refactored my changes a bit to make the changes more obvious.

kamahen commented:
I've reverted my change to sharedparse.py, and added a single __ne__.
I've left my change to the tests, except commenting out some additional
items that this revision to sharedparse.py causes to fail.

If this is OK with you, I'll create a new issue for the magic methods that
don't work and create a new push request that refactors sharedparse.py.

Or would you prefer I do something else?

PS: Sorry for messing things up ... I'm not used to github's workflow.


On 27 March 2018 at 09:53, Guido van Rossum <notifications@github.com>
wrote:

> *@gvanrossum* requested changes on this pull request.
>
> I am really unhappy with this change. I spent 10 minutes writing a
> comparison between the old and new contents of each set and found several
> serious problems. We should not do this refactor and instead go for the
> smallest diff possible (adding __ne__ somewhere).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/mypy/pull/4770#pullrequestreview-107377855>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhBmtYOB-R5L6pxhOorvbUbv1xj7QBfaks5tim6QgaJpZM4SyfyD>
> .
>

gvanrossum commented on test-data/unit/check-warnings.test:
```diff
@@ -145,8 +145,103 @@ main:4: warning: Returning Any from function declared to return "int"
+            return bool(self.x != other.x)
+        else:
+            return NotImplemented
+# TODO: If the programmer specifies "-> Union[bool, type(NotImplemented)]",
```
The need for all these comments here still makes me unhappy, and nothing seems to answer the question why `__ne__` is special.

gvanrossum commented on test-data/unit/check-warnings.test:
```diff
@@ -145,8 +145,103 @@ main:4: warning: Returning Any from function declared to return "int"
+from typing import Any
 class A:
-    def __eq__(self, other: object) -> bool: return NotImplemented
+    # The commented-out items give a warning: Returning Any from function declared to return "A" 
```
Well given the reason why NotImplemented exists that's not surprising. Do any of these do something reasonable at runtime? (Where reasonable == "triggers some default action other than passing through NotImplemented".)

The crashes are interesting though -- please report those separately.

kamahen commented:
I've removed all the commented-out tests ... only the tests that pass are there now. (I'll do separate PRs for the refactoring and the tests that aren't passing).

For looking at the validity of `NotImplemented`, I'd need to inspect the inner workings of both mypy and the Python interpreter (the documentation is a bit hand-wavy on the details, although if I think about it hard enough, I think the answer about whether `NotImplemented` makes sense or not for an individual `__` method is "obvious"). One additional question ... is it intended that `NotImplemeted` be only used by `__` methods that correspond to operators, or should it be available for general use?

gvanrossum commented:
Only for binary operators -- its purpose is exclusively to make the
interpreter try __radd__ and friends.

https://docs.python.org/3/library/constants.html#NotImplemented

On Tue, Mar 27, 2018 at 3:58 PM, Peter Ludemann <notifications@github.com>
wrote:

> I've removed all the commented-out tests ... only the tests that pass are
> there now. (I'll do separate PRs for the refactoring and the tests that
> aren't passing).
>
> For looking at the validity of NotImplemented, I'd need to inspect the
> inner workings of both mypy and the Python interpreter (the documentation
> is a bit hand-wavy on the details, although if I think about it hard
> enough, I think the answer about whether NotImplemented makes sense or
> not for an individual __ method is "obvious"). One additional question
> ... is it intended that NotImplemeted be only used by __ methods that
> correspond to operators, or should it be available for general use?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/mypy/pull/4770#issuecomment-376702732>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACwrMrcQxBTI5UOjdglAqIzpJFLasvyzks5tisQDgaJpZM4SyfyD>
> .
>



-- 
--Guido van Rossum (python.org/~guido)

kamahen commented:
On 27 March 2018 at 16:03, Guido van Rossum <notifications@github.com>
wrote:

> Only for binary operators -- its purpose is exclusively to make the
> interpreter try __radd__ and friends.
>
> https://docs.python.org/3/library/constants.html#NotImplemented
>
>
​
Ahh, I knew it must be documented somewhere. ;)

Anyway, is the rest of this PR acceptable now? (Based on the documentation
for `NotImplemented`, there are some lingering issues for other magic
methods, but I'll deal with them separately)
​

gvanrossum commented on test-data/unit/check-warnings.test:
```diff
@@ -145,8 +145,70 @@ main:4: warning: Returning Any from function declared to return "int"
 # flags: --warn-return-any
+from typing import Any
+class A:
+    def __init__(self, x: Any) -> None:
```
Actually the tests in this file are specifically for the `--warn-*-any` flags, and this particular test just tests that if a binary magic method returns `NotImplemented` you don't get a warning about returning `Any` (which apparently requires some effort since `NotImplemented` is declared as having type `Any` in typeshed as well as in test-data/unit/fixtures/notimplemented.pyi.

Tests that all binary methods support returning `NotImplemented` (or that other methods don't!) belong somewhere else, and tests that mypy infers the right return type when various forms of branching are present belong elsewhere again.

I trust that the branching logic is very well tested already, so I don't think we need more tests for that. As to where the basic `return NotImplemented` tests should go, I think the best place is probably `check-classes.test`. This is somewhat arbitrary.

kamahen commented:
It's now a 1-line fix ... I'll deal with the other issues separately. Some of them may not be worth the price of "improvement". ;)
